sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -2488,59 +2539,117 @@ case class Encode(value: Expression, charset: Expression)
   """,
   since = "3.3.0",
   group = "string_funcs")
-// scalastyle:on line.size.limit
-case class ToBinary(
-    expr: Expression,
-    format: Option[Expression],
-    nullOnInvalidFormat: Boolean = false) extends RuntimeReplaceable
-  with ImplicitCastInputTypes {
-
-  override lazy val replacement: Expression = format.map { f =>
-    assert(f.foldable && (f.dataType == StringType || f.dataType == NullType))
```

Comment:
Note that, I'm asking for justification because this is not free. We changed some logic from compile time to runtime, which complicates the implementation and may also affect performance.

Reply:
@cloud-fan Here is my attempt to revert to `RuntimeReplaceable`. It works for the most part except for errors. I am not sure if it is serialization to executors or something else but `QueryContext` is pretty much empty (not `None`, just all fields are not populated). <- @MaxGekk is there a way to fix this? 

Another alternative way to enforce `format` being foldable/literal is to check it in `checkInputDataTypes`. This will revert my last commit but error messages will be cleaner, what do you think?